### PR TITLE
Extend featurizer

### LINKED ISSUE #118: Orbital resolved data in featurizer
Add possibility to obtain orbital-resolved features in featurizer . Should be useful to get data for orbital interaction diagrams as well

- [x] COXX fingerprints
- [x] COXX center, width, skew, kurtosis

### LINKED ISSUE #143: Run some more tests on the features beyond our Database paper
**Describe the bug**
I think we would need to test these features more (e.g., test on compounds where we know what should come out)

### PR DESCRIPTION
**Pull request by naik-aakash**
Closes #118, #143, #145 and #183 

# Todo
- [x] include summary stats from orbital resolved analysis as features (is Optional)
- [x] make coxx moment features calc methods static
- [x] add coxx_edge (max peak position) as features
- [x] adapt moment calc methods in FeaturizeCOXX also to take in orbitals (makes it possible to obtain fingerprints of particular orbital / moment features for orbital resolved coxx)
- [x] Add batch structure graph Featurizer
- [x] Add support to batchsummary featurizer for newly formatted cba jsons
- [x] Add LOBSTER DOS featurizer
- [x] Add test file for new cba jsons
- [x] Add optional element dos moment as features to dos featurizer
- [x] Better handle LobsterPy  BatchFeaturizer for the case when no bonds are detected in `all` bonds mode. Currently it stops the script. This is not expected behaviour(Observed for Nobel gases  or dimers)
- [x] Add tests for extended capabilities (Will be added after pytest PR is merged to avoid rewriting more tests/ fixing documentation)
- [x] Add tests using existing other test data (should close #143 )
- [x] Remove setup method from structuregraph tests (prolongs test execution duration)

### CODE REVIEW

**Comment by naik-aakash** [comment]
Hi @JaGeo, This PR can be reviewed and merged.  I think this is mainly one of the last new feature additions PR

I will raise another PR to increase the coverage of the plotter and CLI modules

**Comment by JaGeo** [comment]
@naik-aakash Could you document how to use it and add example outputs here? It makes the review much easier.

**Comment by naik-aakash** [comment]
I will add the documentation on the usage in another PR. It makes this harder to do in this PR; as complete documentation is reworked in another PR, then it will be repetitive work if I try to add all this here in an old format and then transfer it to the new format (also more merge conflicts to solve )

I can send you a notebook demonstrating the usage and expected outputs would this help?

**Comment by JaGeo** [comment]
Could you please demonstrate it here? I would need to install yet another Python environment to review it.

**Comment by naik-aakash** [comment]
# Using DOS featurizer
image

**Comment by naik-aakash** [comment]
image

**Comment by naik-aakash** [comment]
# Using static method to get moment features for orbitals from completel cohp obj and label list
image


`Note:` this is intended just for our internal usage so no tests are implement for this as you will see in coverage as well. And it is currently not used as features in our featurizers , i.e the moment features using orbitals

I implemented them just for the need in case we decide to go for orbital interaction diagrams in the near future. Thus method is also not public in the class.

**Comment by naik-aakash** [comment]
Hi @JaGeo, I think that's all new things I added. The rest of the changes are minor code tweaks.

**Comment by JaGeo** [comment]
> # Using static method to get moment features for orbitals from completel cohp obj and label list
> image
> 
> `Note:` this is intended just for our internal usage so no tests are implement for this as you will see in coverage as well. And it is currently not used as features in our featurizers , i.e the moment features using orbitals
> 
> I implemented them just for the need in case we decide to go for orbital interaction diagrams in the near future. Thus method is also not public in the class.

How much effort would it be to add tests and more documentation? Especially in advance of submitting a paper to JOSS, it would be important to test as much code as possible.

**Comment by naik-aakash** [comment]
> > # Using static method to get moment features for orbitals from completel cohp obj and label list
> > image
> > `Note:` this is intended just for our internal usage so no tests are implement for this as you will see in coverage as well. And it is currently not used as features in our featurizers , i.e the moment features using orbitals
> > I implemented them just for the need in case we decide to go for orbital interaction diagrams in the near future. Thus method is also not public in the class.
> 
> How much effort would it be to add tests and more documentation? Especially in advance of submitting a paper to JOSS, it would be important to test as much code as possible.

Not much to be honest. As I had shown you the updated documentation, Just few more things to need to be add in that after this is merged. I can add the test for this specific things as well when working on increasing coverage for cli and plotting module. We should be able to submit it to JOSS I guess after I finish updating the documentation and text of the paper.

**Comment by naik-aakash** [comment]
And Incase if I was not clear, only test for orbital moment features are not added (internal usage), for all the rest of the things, tests are already in place for DOS featurizer, Structure graphs etc.

**Comment by naik-aakash** [comment]
@JaGeo, I have added tests for cohp orbital moment features now. I think should be ready for review now

File: lobsterpy/cli.py [hunk 1]
```diff
@@ -915,8 +915,17 @@ def run(args):
                 filename = filename.with_name(zpath(filename.name))
             options = {"are_cobis": False, "are_coops": False}
 
+        struture_filename = args.structure.parent / "POSCAR"
```
Comment (JaGeo) [inline lobsterpy/cli.py:918]: Can we have a fallback option "POSCAR.lobster". I think the latest Lobster supports this now.

File: lobsterpy/featurize/batch.py [hunk 2]
```diff
@@ -671,19 +672,346 @@ def _get_fingerprints_df(self) -> pd.DataFrame:
...
+
+        df = pd.DataFrame(index=[ids])
+
+        df.loc[ids, "structure_graph"] = graph.sg
+
+        return df
+
+    def get_df(self) -> pd.DataFrame:
+        """
+        Generate a pandas dataframe with structure graph with LOBSTER data.
```
Comment (JaGeo) [inline lobsterpy/featurize/batch.py:791]: Does one need a dataframe for typical applications of the structure graphs?

File: lobsterpy/featurize/core.py [hunk 3]
```diff
@@ -528,9 +725,8 @@ def _calculate_wicoxx_ein(self) -> tuple[float, float, float, float]:
             icoxx = self.icoxxlist.icohpcollection.get_icohp_by_label(
                 lab, summed_spin_channels=True
             )
-            weight = (
-                icoxx / icoxx_total
-            )  # calculate the weights based on icohp contri to total icohp of the structure
+            # calculate the weights based on icohp contri to total icohp of the structure
+            weight = (icoxx / icoxx_total) if icoxx_total != 0 else 0
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:729]: I don't think this is safe. Please make sure you really cutoff numerical issues. Please compare the floats correctly. We had this issue multiple times before.

File: lobsterpy/featurize/core.py [hunk 4]
```diff
@@ -570,39 +766,58 @@ def _calculate_wicoxx_ein(self) -> tuple[float, float, float, float]:
...
             label_list: List of bond labels
-            per_bond: Will scale cohp values by number of bonds i.e .length of label_list arg
+            orbital: orbital for which moment features need to be calculated. Cannot be used independently.
+            Always needs a label_list.
+            per_bond: Will scale cohp values by number of bonds i.e. length of label_list arg
             (Only affects when label_list is not None)
 
         Returns:
-            coxx center,width, skewness, and kurtosis in eV
+            coxx center,width, skewness, kurtosis and edge in eV
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:792]:
```suggestion
            coxx center, width, skewness, kurtosis, and edge in eV
```

File: lobsterpy/featurize/core.py [hunk 5]
```diff
@@ -620,57 +835,60 @@ def _calc_moment_features(
...
 
-        return coxx_center, coxx_width, coxx_skew, coxx_kurt
+        return coxx_center, coxx_width, coxx_skew, coxx_kurt, coxx_edge
 
-    def _get_coxx_center(
-        self,
+    @staticmethod
+    def get_coxx_center(
         coxx: npt.NDArray[np.floating],
         energies: npt.NDArray[np.floating],
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:893]: what type is this?

File: lobsterpy/featurize/core.py [hunk 6]
```diff
@@ -873,7 +1119,11 @@ def _calc_ionicity(self) -> float:
                 )
             ):
                 valence_elec = element(structure.species[i].value)
-                val = j / (valence_elec.nvalence() - 0)
+                val = (
+                    j / (valence_elec.nvalence() - 0)
+                    if valence_elec.nvalence() != tol
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:1124]: Could you check this here as well? I am also not sure what nvalence is and whether != is really a good way to compare floats or integers.

**Review by JaGeo — APPROVED** [review]
I found some numerically questionable comparisons. Could you check those and check if you did something similar elsewhere? I feel != is partially quite brittle.

**Review by JaGeo — CHANGES_REQUESTED** [review]
Please change the raised points :).

File: lobsterpy/featurize/batch.py (hunk 2 above)
Comment (naik-aakash) [reply lobsterpy/featurize/batch.py:791]: One would not typically need dataframe, I want to keep all the featurizers consistent and having all graphs returned in a single dataframe makes it easier to use them further down the line

Comment (JaGeo) [reply lobsterpy/featurize/batch.py:791]: Okay, good!

File: lobsterpy/featurize/core.py (hunk 6 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:1124]: it is integer number always. I want to avoid zero divison error here

Comment (JaGeo) [reply lobsterpy/featurize/core.py:1124]: But what is "tol"? This sounds like a float

Comment (JaGeo) [reply lobsterpy/featurize/core.py:1124]: Also line before: why -0?

File: lobsterpy/featurize/core.py (hunk 5 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:893]: Numpy array , will change it to usual notation

Comment (JaGeo) [reply lobsterpy/featurize/core.py:893]: Npt is confusing

File: lobsterpy/featurize/core.py (hunk 6 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:1124]: Sorry about that dumb mistake, -0 does not make sense. I fixed it now using explicit try except block

**Comment by naik-aakash** [comment]
> Please change the raised points :).

Done 😄

File: lobsterpy/featurize/core.py (hunk 3 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py]: Fixed it using explicit try-except block

File: lobsterpy/cli.py (hunk 1 above)
Comment (naik-aakash) [reply lobsterpy/cli.py:918]: I think it is better if we do this in another PR, we do not yet support the newer LOBSTER version or also tested it. But it is good idea though!

**Review by JaGeo — APPROVED** [review]

**Comment by JaGeo** [comment]
Merged but please showcase how it works in the updated documentation

**Comment by naik-aakash** [comment]
> Merged but please showcase how it works in the updated documentation

Thanks! Yes, I am on it 👍

**Pull request merged**